File: patterns/java/src/refactoring_guru/patterns/strategy/internet_order/strategies/PayByCreditCard.java
package refactoring_guru.patterns.strategy.internet_order.strategies;

import java.io.*;

/**
 * RU: Конкретная стратегия, реализует оплату корзины интернет магазина
 * кредитной картой клиента.
 */
public class PayByCreditCard implements PayStrategy {
    private final BufferedReader READER = new BufferedReader(new InputStreamReader(System.in));
    private CreditCard card;

    /**
     * RU: Проверяем карту клиента.
     */
    @Override
    public void collectPaymentDetails() {
        try {
            System.out.print("Enter card number: ");
            String number = READER.readLine();
            System.out.print("Enter date 'mm/yy': ");
            String date = READER.readLine();
            System.out.print("Enter cvv code: ");
            String cvv = READER.readLine();
            card = new CreditCard(number, date, cvv);

            // RU: Валидируем номер карты.

        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    /**
     * RU: После проверки карты мы можем оплатить покупку,
     * если клиент продолжает покупки, мы не запрашиваем карту заново.
     */
    @Override
    public boolean pay(int paymentAmount) {
        if (cardIsPresent()) {
            System.out.println("Paying " + paymentAmount + " using Credit Card");
            card.setAmount(card.getAmount() - paymentAmount);
            return true;
        } else {
            return false;
        }
    }

    private boolean cardIsPresent() {
        return card != null;
    }
}

File: patterns/cs/src/RefactoringGuru/Strategy/Example/Strategies/PayByCreditCard.cs
using System;

namespace RefactoringGuru.Strategy.Example.Strategies
{
    /**
     * EN: Concrete strategy. Implements credit card payment method.
     * 
     * RU: Конкретная стратегия. Реализует оплату корзины интернет магазина
     * кредитной картой клиента.
     */
    class PayByCreditCard : PayStrategy
    {
        private CreditCard card;

        /**
         * EN: Collect credit card data.
         * 
         * RU: Собираем данные карты клиента.
         */
        public void CollectPaymentDetails()
        {
            Console.WriteLine("Enter card number: ");
            String number = Console.ReadLine();
            Console.WriteLine("Enter date 'mm/yy': ");
            String date = Console.ReadLine();
            Console.WriteLine("Enter cvv code: ");
            String cvv = Console.ReadLine();
            card = new CreditCard(number, date, cvv);

            // EN: Validate the card number.
            // 
            // RU: Валидируем номер карты.
        }

        /**
         * EN: After card validation we can charge customer's credit card.
         * 
         * RU: После проверки карты мы можем совершить оплату. Если клиент
         * продолжает покупки, мы не запрашиваем карту заново.
         */
        public bool Pay(int paymentAmount)
        {
            if (CardIsPresent())
            {
                Console.WriteLine("Paying " + paymentAmount + " using Credit Card");
                card.SetAmount(card.GetAmount() - paymentAmount);
                return true;
            }
            else
            {
                return false;
            }
        }

        private bool CardIsPresent()
        {
            return card != null;
        }

    }
}
